### PR TITLE
ENH: Use SymEngine in ESPEI

### PR DESCRIPTION
**Pull request by richardotis**
This PR depends on https://github.com/pycalphad/pycalphad/pull/376 - tests will fail without it

Similar to the ongoing work in the above PR, this one creates a number of compatibility fixes so that ESPEI can continue to work with the fully SymEngine-based pycalphad. It goes further to eliminate SymPy as a dependency entirely by replacing all SymPy based objects with SymEngine equivalents. The test suite passes locally when used with the above pycalphad PR. There is a slight performance difference: roughly **25 seconds** on this PR, versus **50 seconds** running ESPEI master and pycalphad develop. I expect this difference to be more discernible in more complex cases, especially with regard to the cold start time.

As with pycalphad, there are no direct changes to user-facing APIs here, but direct manipulation of SymEngine objects will have different semantics, so code that isn't covered by the suite may have breakages. We'll need to "kick the tires" here to convince ourselves that we're getting deterministic and numerically correct results.

Note: I temporarily modified the 'pycalphad develop' test to use a commit from https://github.com/pycalphad/pycalphad/pull/376 for demonstration purposes

### CODE REVIEW

File: tests/test_parameter_generation.py [hunk 1]
```diff
@@ -439,10 +439,12 @@ def test_cpm_sm_data_can_be_fit_successively(datasets_db):
...
     # beware that the calculate() results will not match up exactly with the original data due to rounding of parameters
-    assert dbf.symbols['VV0000'] == 105.255  # T*ln(T) L2 term
-    assert dbf.symbols['VV0001'] == -40.953  # T*ln(T) L1 term
-    assert dbf.symbols['VV0002'] == -44.57 # T*ln(T) L0 term
-    assert dbf.symbols['VV0003'] == 36.6556 # L0 T term, found after CPM_MIX addition
+
+    assert dbf.symbols['VV0000'] == 36.6556  # L0 T term, found after CPM_MIX addition
+    assert dbf.symbols['VV0001'] == 105.255  # T*ln(T) L2 term
+    assert dbf.symbols['VV0002'] == -40.953  # T*ln(T) L1 term
+    assert dbf.symbols['VV0003'] == -44.57 # T*ln(T) L0 term
```
Comment (bocklund) [inline tests/test_parameter_generation.py:446]: @richardotis are the ordering changes here due to a change in the sorting of T-dependent terms from the pycalphad side?

Comment (richardotis) [reply tests/test_parameter_generation.py:446]: Yes. The terms were ordered by their string representation, and it appears that SymEngine orders the terms slightly differently from SymPy.

Comment (bocklund) [reply tests/test_parameter_generation.py:446]: My preference is for this ordering (it was always a nit of mine in the way they were sorted). This may be a breaking corner case in some existing workflows that depend on a particular ordering.

I don't think it blocks this issue. I might actually use the breaking change as an opportunity to reorder the numbering to go from L0...L4 rather than L4...L0.

Comment (richardotis) [reply tests/test_parameter_generation.py:446]: This order could be controlled directly by writing a sorting function that explicitly sets the exponent of each variable. In principle this change breaks the determinism of any existing end-to-end workflow, though previously generated TDBs should give the same result in the MCMC step. It depends on what sort of guarantees you're looking to provide here.

**Comment by bocklund** [comment]
Locally, the master branch merges cleanly onto this one. It should just need a merge, then bump symengine version. If all goes well (as it does locally for me), then we should just need to bump the pycalphad pin when a new pycalphad version is released

**Comment by bocklund** [comment]
I merged the master branch, bumped the pycalphad test hash, and added a new failing test that I discovered.

File: espei/paramselect.py [hunk 2]
```diff
@@ -422,15 +417,15 @@ def phase_fit(dbf, phase_name, symmetry, datasets, refdata, ridge_alpha, aicc_pe
             check_symbol = Symbol('YS') * Symbol('Z')**degree
             keys_to_remove = []
             for key, value in sorted(parameters.items(), key=str):
-                if key.has(check_symbol):
+                if key.xreplace({check_symbol: 0}) != key:
```
Comment (bocklund) [inline espei/paramselect.py:420]: In manual testing, I found that this workaround for `has` doesn't work for interactions with temperature terms. I added a test (`test_parameter_generation.test_high_order_interaction_terms_no_spurious_symbols`) to exercise the bug.

The output from that test contains

```python
TRACE:espei.paramselect - Polynomial coefs: [T*VV0000 + T*Z*VV0001, 0.0, 0.0, 0.0, 0, 0, 0, 0, 0, 0]
```

i.e. what's happening is that the `YS*Z` feature doesn't get xreplaced when `degree = 1`, but the `YS` (constant T) feature that becomes V0000 gets replaced. 

Here's some code to reproduce:

```python
# Uncomment one import or the other to test
# from symengine import Symbol
# from sympy import Symbol

key = Symbol("YS") * Symbol("Z")
key_with_T = Symbol("T") * Symbol("YS") * Symbol("Z")
check_symbol = Symbol("YS") * Symbol("Z")
print("check_symbol", check_symbol)
# Sympy and SymEngine both give the same results here:
# When key == check_symbol, the xreplace works; when T is multiplied, the xreplace doesn't replace
print("key", key)
print("key replaced", key.xreplace({check_symbol: 0}))
print("key_with_T", key_with_T)
print("key_with_T replaced", key_with_T.xreplace({check_symbol: 0}))

# Only SymPy works (gives `True`), SymEngine raises with
# TypeError: <class 'symengine.lib.symengine_wrapper.Symbol'> required. Z*YS is of type <class 'symengine.lib.symengine_wrapper.Mul'>
print(key.has(check_symbol))
```

Comment (richardotis) [reply espei/paramselect.py:420]: I think this is https://github.com/symengine/symengine/issues/1359

Comment (richardotis) [reply espei/paramselect.py:420]: This function will pass the test you made:

```python
def has_symbol(expr, check_symbol):
    """
    Workaround for SymEngine not supporting Basic.has() with non-Symbol arguments.
    Only works for detecting subsets of multiplication of variables.
    """
    try:
        return expr.has(check_symbol)
    except TypeError:
        for at in expr.atoms(symengine.Mul):
            if set(check_symbol.args).issubset(at.args):
                return True
    return False
```

However, it will fail when dealing with powers of variables in the expression:
```python
has_symbol(Symbol("YS") * (Symbol("Z") ** 2), Symbol("YS") * (Symbol("Z")))
# returns False, but is actually True
````

Comment (bocklund) [reply espei/paramselect.py:420]: The purpose of this chunk of code is to associate the symbolic representation of a feature `T*YS*Z` to the RK interaction order (`degree`) so it can be entered into the database. We go in reverse order so that the high order terms are matched and removed first. 

This could be a good start as a workaround if we extend it for powers. There might be some other changes we could make to accomplish the same thing

Comment (richardotis) [reply espei/paramselect.py:420]: This seems to work:
```python
def _poly_degrees(expr):
    poly_dict = {}
    for at in expr.atoms(symengine.Symbol):
        poly_dict[at] = 1
    for at in expr.atoms(symengine.log):
        poly_dict[at] = 1
    for at in expr.atoms(symengine.Pow):
        poly_dict[at.args[0]] = at.args[1]
    return poly_dict

def has_symbol(expr, check_symbol):
    """
    Workaround for SymEngine not supporting Basic.has() with non-Symbol arguments.
    Only works for detecting subsets of multiplication of variables.
    """
    try:
        return expr.has(check_symbol)
    except TypeError:
        expr_poly = _poly_degrees(expr)
        check_poly = _poly_degrees(check_symbol)
        print(expr, check_symbol, expr_poly, check_poly)
        for cs, check_degree in check_poly.items():
            eps = expr_poly.get(cs, 0)
            if check_degree > eps:
                return False
        return True
```

File: tests/test_parameter_generation.py (hunk 1 above)
Comment (richardotis) [reply tests/test_parameter_generation.py]: I wrote a sorting function (`stable_sort_key`) that seems to reproduce the old ordering, so I reverted the changes to the test.

File: .github/workflows/tests.yaml [hunk 3]
```diff
@@ -61,7 +61,7 @@ jobs:
       with:
         python-version: ${{ matrix.python-version }}
     - name: Install pycalphad development version
-      run: pip install git+https://github.com/pycalphad/pycalphad.git@develop
+      run: pip install git+https://github.com/pycalphad/pycalphad.git@9a12344085dd0ba720e57c1c5ebafbf26a5d0c30
```
Comment (bocklund) [inline .github/workflows/tests.yaml:64]: revert
```suggestion
      run: pip install git+https://github.com/pycalphad/pycalphad.git@develop
```

**Review by bocklund — APPROVED** [review]
